icu4c/source/test/intltest/rbbitst.cpp

```diff
@@ -5231,7 +5231,7 @@ void RBBITest::TestTraceCreateBreakEngine(void) {
 
     // To word break the following text, BreakIterator will create 5 dictionary
     // break engine internally.
-    brkitr->setText(
+    UnicodeString text(
```

Comment:
Note for reviewers: this change fixes a pre-existing bug that began causing CI failures with this PR. With the original code, brkitr->setText(UnicodeString &s) was being passed a temporary string that immediately went out of scope. But setText() retains a reference to it for use during iteration, which becomes a dangling reference.

Reply:
lgtm but I am curious why this is failing only now.
Does this have something to do with the switch to a UCharCharacterIterator?

Reply:
Good question, I was curious too. It does indeed have to do with the switch to UCharCharIterator.

StringCharacterIterator::setText() makes a copy of the incoming UnicodeString.
Clever reference counted buffer sharing in UnicodeString ends up keeping the heap allocated buffer alive after the original UnicodeString goes out of scope, for use by the copied string. And it is that buffer that the break iterator operates on, via UText.

The original code would still fail (I think) if the UnicodeString was short enough to use the string's internal buffer instead of the heap buffer.

It's all kind of funny, that the CharacterIterator stuff is only there for compatibility with obsolete API functions that probably no one is using, but its implementation has this unintended side effect on fundamental rbbi APIs that everyone uses.

There's a real chance that this change will break some existing users. What, if anything, to do about it warrants further discussion. Copy the input string all of the time, because that's the least error prone? Never copy, because that's what we've always said we do? Copy some of the time, to avoid potential new breakages? (But what would we be promising, and how would we ever explain it?)

No matter what we decide, I'd like to separate the question from this PR, and from this Jira issue, so that it gets better visibility.

Reply:
Ok, so we have done more than we promise:
> The BreakIterator will retain a reference to the supplied string. The caller must not modify or delete the text while the BreakIterator retains the reference.

What we say is consistent with other APIs, so no surprise. And at least some of our users will be calling the C API which can only take pointer+length.